Clean slate, initial implementation with tests

RangerMauve opened the pull request.

RangerMauve committed:
Clean slate, initial implementation with tests

RangerMauve committed:
Delete package-lock

RangerMauve committed:
Add encryptionKey to OSM, garbage collect mapeo instances after a while

RangerMauve committed:
Refactor, track valid mapeo instances, add CLI

gmaclennan commented:
@RangerMauve I'm just looking through this, can you ensure that the tests actually test that sync works (e.g. that a document and an image stored on each peer are available on the other peer after sync)? Reading through the code it looks to me like `replicateFromWebsocket()` will connect to a peer, but will not initiate sync. The code in mapeo-core is in dire need of a re-write, but the command that actually kicks off sync is [`peer.handshake.accept()`](https://github.com/digidem/mapeo-core/blob/ws-sync/sync.js#L306). This property (`peer.handshake`) is added by the [`addPeer()`](https://github.com/digidem/mapeo-core/blob/ws-sync/sync.js#L568) method, and I don't think sync starts until the handshake is accepted. This is flakey code anyway, because `peer.handshake` is added and then removed, and if the user tries to sync when it is `undefined`, then sync fails.

RangerMauve commented:
Sweet, thank you for the input. I'll make sure to take that into account.

Would creating data on one side and trying to load it on the other side after sync suffice?

gmaclennan commented:
Yes, see the sync tests in mapeo-core for some exampes about how to do this: https://github.com/digidem/mapeo-core/blob/ws-sync/test/sync.js#L214

RangerMauve committed:
Refactor to ensure replication is happening

RangerMauve committed:
Actually test the data got replicated over

RangerMauve committed:
Use bin in package.json

RangerMauve committed:
Add docs for setting up service

RangerMauve committed:
Add user and group to mapeo-web service for security

RangerMauve committed:
Use new connectWebsocket API

RangerMauve committed:
Fix passing down encryptionKey

RangerMauve committed:
Set deviceType to 'cloud'

RangerMauve committed:
Use discoveryKey instead of projectKey in replication

gmaclennan commented on README.md:
```diff
@@ -39,6 +39,68 @@ Mapeo Web allows project owners to generate special GeoJSON export URLs. These U
+  server_name cloud.mapeo.app;
+
+  location / {
+    proxy_pass http://localhost:62736;
```
What's the default timeout on http connection for nginx? Sync could take 30 mins or more over a slow connection, so we want to ensure nginx would not timeout.

RangerMauve replied:
Good call, what should I set it to?

Right now it looks like the timeout is there for when there are no bytes being sent or received over a 60 second interval.

If the sync is happening for 30 mins or more and is actually sending data during that time, the default config should work.

https://developpaper.com/detailed-explanation-of-nginx-timeout-configuration/

```
server {
   ...
   proxy_read_timeout 300;
   proxy_connect_timeout 300;
   proxy_send_timeout 300;
   ...
}
```

gmaclennan replied:
ok good. Just to be sure, I think a 5 minute keep-alive is good. I have seen satellite internet with > 60 second ping times. I think by the time you get to 5 minutes without data, you may as well give up trying to sync!

gmaclennan commented on bin.js:
```diff
@@ -0,0 +1,44 @@
+    default: DEFAULT_NAME
+  })
+  .option('gc-delay', {
+    describe: 'Delay before garbage collecting Mapeo instances',
```
Nitpik, isn't this the delay until instances are released for gc, but gc could happen much later?

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,85 @@
-  var dbdir = path.join('projects', pid, 'db')
-  var osm = Osm({ dir: dbdir, encryptionKey: pid })
-  var media = Blob(path.join('projects', pid, 'media'))
+    this.fastify.put('/permissions/project/:key', (req, reply) => {
```
I think we should treat this as the idea of creating a "project" record on the server. We may want to add more refined permissions in the future. I think it would also be good practice to not expose the project key on any public end point, e.g. key is write only, projectId === discovery key and is used for public transactions.

Therefore: path could be `/projects/` (plural to follow common REST semantics)
Create a new project on the server, POST to `/projects/` with `key` in body.
Returns project ID (=== discovery key)
Delete a project, DELETE `/projects/:discoveryKey`.

Initially, existence of project record === permission to sync.
In the future: might add more details to a project record to indicate refined permissions (e.g. resource limits).

RangerMauve replied:
Is having the ID in the response necessary?

gmaclennan replied:
If we build a front-end "admin" UI for this (to add projects to Mapeo web and configure their permissions), then it is useful get back the ID when you create a new object. The client could assume the ID of a newly created project resource would be the discovery key, which the client could also derive, but if we return the ID then we don't hold ourselves to that (a project ID could be something different in the future)

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,135 @@
+  get (key) {
+    if (this.instances.has(key)) return this.instances.get(key)
+    const { storageLocation, id } = this
+    const dir = path.join(storageLocation, 'instances', key)
```
Let's use the discovery key for this. I want to avoid leaving secret keys exposed where possible.

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,135 @@
+    })
+
+    mapeo.sync.on('peer', (peer) => {
+    // TODO: Should we track / report sync progress somewhere?
```
It would be good, especially during our initial piloting of this service, to log all sync progress somewhere, so that we can get an idea of how it is working and what performance issues there are. Do you have experience with logging services? The one I know of is https://apex.sh/logs/ but I've never used it.

RangerMauve replied:
I was thinking it'd be to STDOUT and to use journald for managing them. Is using an external logging service a hard requirement?

gmaclennan replied:
No, I think that would be a good solution. We'd have to keep an eye on things, e.g. how long before they could take up loads of disk space? We would need to have our own way of analysing the logs and querying them. I have _zero_ experience in all this!

RangerMauve replied:
Is this something we should have before this PR gets merged, or could we have it as a TODO for after the initial testing?

RangerMauve replied:
I've added some basic logging using fastify and pino, so we should at least see things replicating.

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,135 @@
+    name = DEFAULT_NAME
+  }) {
+    super()
+    // TODO: Add leveldb to track
```
Yes, let's have a single place, e.g. a leveldb, where secret project keys are written, and use the derived discovery key for the identifier in all other places in the app.

RangerMauve replied:
Only caveat is that `multi.get()` needs the project key so it can initialize the mapeo instance properly.

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,135 @@
+      connections.delete(connection)
+      setTimeout(() => {
+        // New connections have been made since the last time
+        if (connections.size()) return
```
I'm worried this logic is a little hard to parse and could result in unexpected behaviour.

Each peer that connects is setting a timeout, and closing an instance will happen if the timeout happens at a moment when there are no peers connected.

I think the expected logic would be: instance is closed gcTimeout seconds after the last peer finished syncing. I don't think that behaviour is guaranteed by this code right now.

RangerMauve replied:
Are there peers that stay connected after they've finished syncing?

gmaclennan replied:
I don't think so. I don't think I was clear, and maybe I am mis-understanding the code here, but this is the scenario I am thinking:

GC_TIMEOUT = 600s

0s: Peer A connects, starts to sync
100s: Peer A finishes sync, disconnects, timeout A is started.
...
500s: Peer B connects, starts sync
650s: Peer B finishes sync, disconnects, timeout B is started.
700s: timeout A completes, no connections, instance closed.

Here, the instance is closed 50s after the last peer finished sync. I would expect the instance to remain active as long as peers continue to connect and sync within the GC_TIMEOUT time frame.

RangerMauve replied:
Ah I understand now, thank you. I think a good approach there would be to clear any existing timeouts before starting a new one.

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  close (cb) {
+    const total = [...this.instances].length
```
```suggestion
    const total = this.instances.size
```
?

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,135 @@
+        if (count === total) cb(lastError)
+      })
+    }
+  }
```
For the all the issues with promises, this kind of code is so much easier as a promise:

```js
  close () {
    const instances = Array.from(this.instances.values())
    return Promise.all(instances.map(mapeo => promisify(mapeo.close))
  }
```

But fine to keep with callback syntax for now. We plan to [switch mapeo-core to using promises](https://github.com/digidem/mapeo-rfc/blob/995ad997ddfcd3c996c8a9e6b4c570f6ee88b777/RFC-003-mapeo-core.md#12-callback-based).

RangerMauve replied:
Yeah, I usually use promises, but didn't want to stray too far from mapeo-core's code style.

gmaclennan commented on permissions.js:
```diff
@@ -0,0 +1,60 @@
+    })
+  }
+
+  hasProjectKey (key, cb) {
```
nitpick: this and below is a `get` not a `has`.

gmaclennan reviewed (changes requested):
This is looking really great. I added a few comments, some nitpicks. For future maintenance, I would really like to reduce how much we pass around the secret keys for projects. Currently we use them both as an identifier (and dir name). I think we should think of the server state being a list of "projects" that are created on the server, identified by their discovery key, and each project currently only has one property (`canSync=true`), which currently is implied: a project exists, then it can sync. Then there should be a single place that stores the secret keys, which can be looked up by project ID (discovery key). The code mainly does this, but it seems like the secret keys leak out in a few places.

gmaclennan commented:
What tasks are remaining to get this to a place where we are ready to pilot it?

RangerMauve commented:
@gmaclennan Here's what I think is left to pilot it:

- Update the code per your review
- Update mapeo-core in my branches of mapeo-desktop and mapeo-mobile
- Double check that it's working
- Generate a new mapeo settings file (currently using the workshop one for testing which is not ideal)
- Get builds of the app for people to install

I'm guessing we could have all that by next week.

RangerMauve committed:
Updated per PR review

RangerMauve committed:
Updated per PR review

RangerMauve commented:
Made some progress on the review, got some questions for clarification for when you have time @gmaclennan

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,135 @@
+  unget (key, cb) {
+    if (!this.instances.has(key)) return process.nextTick(cb)
+    // TODO: Close mapeo instance
+    this.instances.get(key).close((err) => {
```
This could be called twice, because we are not tracking the `closing` state of a mapeo instance. E.g. in line 73, `instances.has(key)` could return true between the `close()` method being called and the callback.
Small window, but a source of weird bugs down the line.

RangerMauve replied:
Good catch

RangerMauve committed:
Changes for PR review

RangerMauve committed:
Add discoveryKey utility

RangerMauve committed:
Add docs on basic auth

RangerMauve committed:
Add client commands to CLI and tests

RangerMauve committed:
Gracefull shutdown, add logging

RangerMauve commented:
Made it a bit more robust / usable.

RangerMauve commented:
@gmaclennan Could I get a final review on this too please? 😁

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
+
+      this.permissions.getProjectKeyForDiscoveryKey(discoveryKey, (err, projectKey) => {
+        if (err || !projectKey) {
+          return connection.end('Invalid Key')
```
How do errors work in Websocket? Is there a way to indicate an error to the client? Should this be JSON so that it can be parsed correctly on the client? How do we handle this error in client code? Do we test for this error (I think it is a likely error to happen)?

RangerMauve replied:
Raw sockets don't have an error type. The most we can do is end the socket connection.

I don't think there's an easy way to detect this error in the protocol, it would lead to an error being emitted in parsing the protocol data on the client side.

@noffle Do you know if the mapeo protocol has a place to handle error messages in the handshake? We could simulate one of those if they exist.

gmaclennan replied:
I'd like to be clear about error handling before we merge this. What will the client do if it tries to sync and it is not "enabled"?
This is also important behaviour to test. Currently we have a test for successful sync. We should have a test for trying to sync when the project is not enabled on the server, to check that it fails correctly.

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
+    this.fastify.register(require('fastify-websocket'))
+
+    // Leaving room for other major minor and patch versions.
+    this.fastify.get('/replicate/v1/:discoveryKey', { websocket: true }, async (connection, req, params) => {
```
We don't need to return anything from this function for a websocket request right? No need for this to be an async function?

RangerMauve replied:
I think I initially had an `await` in there, but that's since been removed. 😁 Good catch.

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
+    this.fastify.post('/projects/', (req, reply) => {
+      const { projectKey } = req.body
+      this.permissions.addProjectKey(projectKey, (err) => {
+        if (err) return reply.send(err)
```
Should set status code, and error should be JSON and documented.

RangerMauve replied:
Documented how? This could be any sort of uncaught exception. 🤔 Should the JSON structure for errors be documented in the README?

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
+      this.permissions.getProjectKeyForDiscoveryKey(discoveryKey, (err, projectKey) => {
+        if (err) {
+          reply.status(404)
+          reply.send(err)
```
JSON error.

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
+          return
+        }
+        this.permissions.removeProjectKey(projectKey, (err) => {
+          if (err) return reply.send(err)
```
JSON error. Set status code.

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
-            var filter = values[0].value
-            cb(null, [core, filter])
+    this.fastify.post('/projects/', (req, reply) => {
+      const { projectKey } = req.body
```
Should validate body, easy to do with fastify with a schema for req.body.

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
+    this.fastify.get('/projects/', (req, reply) => {
+      this.permissions.getProjectKeys((err, keys) => {
+        if (err) {
+          return reply.send(err)
```
JSON error and status code.

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,108 @@
+        if (err) {
+          return reply.send(err)
+        } else {
+          const data = keys.map((projectKey) => ({ projectKey }))
```
This is a dangerous end point to have, it leaks secret keys for everyone. What do we need it for? Could we list discovery keys instead? I feel like project keys should never be exposed once added.

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,162 @@
+
+    const key = discoveryKey(projectKey)
+
+    this.logger.info({ projectKey, discoveryKey: key }, 'Initializing project')
```
Should not log project keys

gmaclennan commented on multi-mapeo.js:
```diff
@@ -0,0 +1,162 @@
+      const timer = setTimeout(() => {
+        // New connections have been made since the last time
+        if (connections.size()) {
+          this.logger.info({ projectKey }, 'Aborting gc, has connections')
```
don't log project key

gmaclennan reviewed (changes requested):
Hi Mauve, I've left a few comments in the code, but there are a few areas I think need to be adjusted before merging this:

1. Errors: I think we should return error status codes (e.g. 500) and the body of an error response should be JSON (I'm not sure what it is now, but looks like the code is sending an error as the message body? How should the client deal with / parse these errors?)
2. GET /projects/ end point, do we need it? Seems dangerous
3. Logging of project keys: We should never log these, we should treat them as a secret.
4. The code in Multi Mapeo essentially uses the project key as an identifier (for the instances and connections Maps). I think it would be better practice to use the discovery key as the "Project ID", and use this for in-memory Maps. The fewer places the project key is in the code, the fewer places we can accidentally leak it to logs or in responses.

RangerMauve commented:
Thanks for the review Gregor!

Here's some comments:

1. Fastify already responds with JSON when you reply with an error code. The docs for that are here: https://www.fastify.io/docs/latest/Reply/#errors
2. The /projects/ path is behind the same authentication as adding and removing projects. It was mostly there so that we could verify that stuff got added and to get a list of stuff that can be removed (as admins). I'm down to remove it, but I assumed that admins could be trusted with this route.
3. Should I log the discovery keys instead? Or maybe the first or last bytes of the project key the way mapeo-desktop and mapeo-mobile display them in the app?
4. I initially set that up that way because we need the project key when initializing a mapeo-core instance. This is also why replicate needs the project key. I see the benefit in using the discovery key in the map and within all the other methods though. Would leaving projectKey for get and replicate and using discovery key everywhere else work?

gmaclennan commented:
> Thanks for the review Gregor!
> 
> Here's some comments:
> 
> 1. Fastify already responds with JSON when you reply with an error code. The docs for that are here: https://www.fastify.io/docs/latest/Reply/#errors

Oh that's great, that simplifies things a lot!

> 2. The /projects/ path is behind the same authentication as adding and removing projects. It was mostly there so that we could verify that stuff got added and to get a list of stuff that can be removed (as admins). I'm down to remove it, but I assumed that admins could be trusted with this route.

Makes sense. I think this should be a list of project ids, and I think we should use the discovery key as the project id. I know only admins can see this, but it means that we have a single point of failure (the basic auth password), that if someone gets access, then they can get access to all Mapeo projects around the world (because they would have all the encryption keys). I don't think even admins should be given an easy way to get everyones encryption keys.

> 3. Should I log the discovery keys instead? Or maybe the first or last bytes of the project key the way mapeo-desktop and mapeo-mobile display them in the app?

Let's use discovery keys for everything.

> 4. I initially set that up that way because we need the project key when initializing a mapeo-core instance. This is also why replicate needs the project key. I see the benefit in using the discovery key in the map and within all the other methods though. Would leaving projectKey for get and replicate and using discovery key everywhere else work?

Yes.

RangerMauve committed:
Changes per review

RangerMauve committed:
Move cross-fetch dependency

RangerMauve committed:
Add schema validation to body for POST

gmaclennan commented on index.js:
```diff
@@ -1,135 +1,122 @@
+      },
+      handler: (req, reply) => {
+        const { projectKey } = req.body
+        this.permissions.addProjectKey(projectKey, (err) => {
```
Need to check that the project key is valid hex (`/[^0-9a-f]/.test(projectKey)`) because otherwise the server will crash with an uncaught error (because we do not catch the error that can be thrown in the discoveryKey function).

RangerMauve replied:
The error already throws properly, so I'm not sure if anything needs to be added here.

gmaclennan reviewed (changes requested):
Hi Mauve, thanks for these changes, all looking good. I had neglected to review the testing strategy last time, and I think it's important we test some key functionality before deploying this:

1. `test('Adding invalid project key (64 non-hex characters) fails gracefully')`
2. `test('Trying to sync with project key without permission fails')`
3. `test('Mapeo instance closes after timeout from last sync completion')`
4. `test('Mapeo instance does not close if a new sync starts after first completes but before GC timeout')`

RangerMauve committed:
Added tests per review

RangerMauve commented:
I've added the extra unit tests you mentioned, Gregor. It seems that the behavior right now is that the connection closes before any sort of handshake can be done, so I'm not sure what else we could add to it.

Does this look good?

gmaclennan committed:
fix: Fix tests by increasing timeout

Very strange behaviour here on MacOS: the close happens 1ms after
the fail timer from 1000ms to 2010ms, then above 2010ms the test will
pass. Setting to 3000ms for now and hopefully will be fine.

gmaclennan committed:
chore: Add Github CI

gmaclennan committed:
chore: Add CI

gmaclennan committed:
Merge branch 'master' into ws-rewrite

gmaclennan committed:
chore: Add npm-shrinkwrap.json

gmaclennan committed:
chore: https for github deps, shrinkwrap v1

gmaclennan reviewed (approved):
Ok, I think this is good to go now.

gmaclennan committed:
chore: Update mapeo-core to fdcfd3e

gmaclennan commented:
@mauve currently when I run `./bin.js list https://XXX:XXX@cloud.mapeo.app/` against the server I get a list with two items,  which is a duplicated project key. I would expect each project key to only exist once in the list.

mauve commented:
I did get really confused by this notification :)
